packages/react/use-controllable-state/src/use-controllable-state.tsx

```diff
@@ -1,57 +1,96 @@
 import * as React from 'react';
-import { useCallbackRef } from '@radix-ui/react-use-callback-ref';
+import { useLayoutEffect } from '@radix-ui/react-use-layout-effect';
 
-type UseControllableStateParams<T> = {
-  prop?: T | undefined;
-  defaultProp?: T | undefined;
-  onChange?: (state: T) => void;
-};
+// Prevent bundlers from trying to optimize the import
+const useInsertionEffect: typeof useLayoutEffect =
+  (React as any)[' useInsertionEffect '.trim().toString()] || useLayoutEffect;
+
+type ChangeHandler<T> = (state: T) => void;
+type SetStateFn<T> = React.Dispatch<React.SetStateAction<T>>;
 
-type SetStateFn<T> = (prevState?: T) => T;
+interface UseControllableStateParams<T> {
+  prop?: T | undefined;
+  defaultProp: T;
+  onChange?: ChangeHandler<T>;
+  caller?: string;
+}
 
-function useControllableState<T>({
+export function useControllableState<T>({
   prop,
   defaultProp,
   onChange = () => {},
-}: UseControllableStateParams<T>) {
-  const [uncontrolledProp, setUncontrolledProp] = useUncontrolledState({ defaultProp, onChange });
+  caller,
+}: UseControllableStateParams<T>): [T, SetStateFn<T>] {
+  const [uncontrolledProp, setUncontrolledProp, onChangeRef] = useUncontrolledState({
```

Comment:
Returning the saved callback ref removes the need to double-save it internally, saving a needless effect cycle.